### PR TITLE
Add pgsql integration test

### PR DESCRIPTION
**Pull request by Patrik-Stas**
Test integration between `aries_vcx` and underlying `libindy` + `pgsql` plugin

Signed-off-by: Patrik Stas <patrik.stas@absa.africa>

### CODE REVIEW

File: aries_vcx/tests/pgwallet/wallet_plugin.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+use std::ffi::CString;
```
Comment (mirgee) [inline aries_vcx/tests/pgwallet/wallet_plugin.rs:1]: Seems rather some kind of `libloading_utils.rs`? Again, I would suggest separating general and postgres-specific code.

File: aries_vcx/tests/test_pgsql_plugin.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    pub plugin_init_function: Option<String>,
+    // Wallet storage config for agents wallets
+    pub config: Option<Value>,
+    // Wallet storage credentials for agents wallets
+    pub credentials: Option<Value>,
+}
+
+#[test]
+#[cfg(feature = "test_integration")]
+fn it_should_use_pgsql_wallet_to_provision_cloud_agent() {
```
Comment (mirgee) [inline aries_vcx/tests/test_pgsql_plugin.rs:68]: This kind of "it should" framework for naming tests is interesting (I have seen you do it elsewhere), but not really consistent with other test names in the repo - I would suggest something like `test_provision_cloud_agent_with_pgsql_wallet` for consistency

File: aries_vcx/tests/test_pgsql_plugin.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            },
+            "credentials": {
+              "account": "postgres",
+              "password": "mysecretpassword",
+              "admin_acount": "postgres",
+              "admin_password": "mysecretpassword"
+            },
+            "type": "postgres_storage"
+          }"#;
+    let storage_config: WalletStorageConfig = serde_json::from_str(storage_config).unwrap();
```
Comment (mirgee) [inline aries_vcx/tests/test_pgsql_plugin.rs:85]: Why not bypass serde and just define the struct directly?

File: aries_vcx/tests/test_pgsql_plugin.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    info!("Init pgsql storage: Starting.");
```
Comment (mirgee) [inline aries_vcx/tests/test_pgsql_plugin.rs:86]: We usually don't put logs inside of tests except for integration tests

File: aries_vcx/tests/test_pgsql_plugin.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        agent_seed: None,
+    };
+    create_wallet(&config_wallet).unwrap();
+    open_as_main_wallet(&config_wallet).unwrap();
+    let config_issuer = configure_issuer_wallet(enterprise_seed).unwrap();
+    init_issuer_config(&config_issuer).unwrap();
+    provision_cloud_agent(&config_provision_agent).unwrap();
+    close_main_wallet().unwrap();
+    assert_eq!(4, 4);
+}
```
Comment (mirgee) [inline aries_vcx/tests/test_pgsql_plugin.rs:117]: I wonder whether the test necessarily belongs here. Maybe `aries_vcx::libindy::utils::wallet`? But again, you put the test-only code here...

**Review by mirgee — CHANGES_REQUESTED** [review]

File: aries_vcx/tests/test_pgsql_plugin.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    // based on value of xtype and OS
+    pub plugin_init_function: Option<String>,
+    // Wallet storage config for agents wallets
+    pub config: Option<Value>,
+    // Wallet storage credentials for agents wallets
+    pub credentials: Option<Value>,
+}
+
+#[test]
+#[cfg(feature = "test_integration")]
```
Comment (mirgee) [inline aries_vcx/tests/test_pgsql_plugin.rs:67]: Is `test_integration` the best name? Maybe `plugin_test`, to be consistent?

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -106,7 +107,7 @@ jobs:
           POOL_DOCKERFILE_HASH=${{ hashFiles('ci/indy-pool.dockerfile')}}
           ALPINE_CORE_HASH=${{ hashFiles('ci/alpine_core.dockerfile')}}
 
-          LIBVCX_HASH=${LIBVCX_SOURCE_HASH:0:11}-${LIBVCX_NODE_WRAPPERS_HASH:0:11}-${LIBVCX_NODE_AGENT_HASH:0:11}-${LIBVCX_DOCKERFILE_HASH:0:11}-${AGENCY_CLIENT_HASH:0:11}
+          LIBVCX_HASH=${LIBVCX_SOURCE_HASH:0:11}-${ARIESVCX_SOURCE_HASH:0:11}-${LIBVCX_NODE_WRAPPERS_HASH:0:11}-${LIBVCX_NODE_AGENT_HASH:0:11}-${LIBVCX_DOCKERFILE_HASH:0:11}-${AGENCY_CLIENT_HASH:0:11}
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply .github/workflows/main.yml:110]: I think it's okay, the hash represents changes in libvcx layer and everything underneath

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -518,6 +519,67 @@ jobs:
...
+          docker run --rm -d --name postgres --network host -e POSTGRES_PASSWORD=mysecretpassword postgres:12.1
+          docker run --rm -d --name vcxagency --network host --env-file ci/agency/localhost.env $DOCKER_IMG_NAME_AGENCY
+
+      - name: Run agency+pool integration tests
+        run: |
+          set -x
+          docker run --rm -i --name libvcx --network host -e X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_NO_VENDOR=true $DOCKER_IMG_NAME_LIBVCX \
+                              bash -c '(ls -lah $HOME && ls -lah $HOME/libvcx && cd $HOME/aries_vcx && \
+                              cargo --version && \
+                              rustc --version && \
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply .github/workflows/main.yml:580]: removed

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 3 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs:85]: `WalletStorageConfig` is using `Option<String>` for `config` and `credentials` fields, so it wouldn't look any nicer rn

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 4 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs:86]: This is integration test

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 5 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs:117]: It's kind of the other way arround - our existing organization is somewhat incorrect. Integation tests belong to `test` directory, whilst unit tests should be directly in the files/modules they test.

https://doc.rust-lang.org/book/ch11-03-test-organization.html

> The Rust community thinks about tests in terms of two main categories: unit tests and integration tests. Integration tests are entirely external to your library and use your code in the same way any other external code would, using only the public interface and potentially exercising multiple modules per test.

> To create integration tests, you first need a tests directory.

This pgsql is integration test, so I think the placement is correct.

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 6 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs:67]: Agreed, I'll go for `plugin_test `

File: aries_vcx/tests/pgwallet/wallet_plugin.rs (hunk 1 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/pgwallet/wallet_plugin.rs:1]: Changed to `wallet_plugin_loader`, the libloading code is very tailored for loading wallet plugins

File: aries_vcx/tests/pgwallet/mod.rs [hunk 9]
```diff
@@ -0,0 +1,2 @@
+pub(crate) mod dyn_lib;
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply aries_vcx/tests/pgwallet/mod.rs:1]: The lib-loading code is specific to loading indysdk wallet plugins. The directory is dedicated to testing pgsql integration

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 5 above)
Comment (mirgee) [reply aries_vcx/tests/test_pgsql_plugin.rs:116]: But it is inconsistent until we move all integration tests :)

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 4 above)
Comment (mirgee) [reply aries_vcx/tests/test_pgsql_plugin.rs:85]: OK, let me rephrase that: "We usually don't put logs inside of tests except for end-to-end integration tests"

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 3 above)
Comment (mirgee) [reply aries_vcx/tests/test_pgsql_plugin.rs:84]: That could also be argument for defining the struct directly - the fact that some of the fields are `Option`s would not be hidden anymore. And not using dependencies when not necessary is a good rule of thumb. But that's just my opinion, I will leave it up to you.

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 4 above)
Comment (mirgee) [reply aries_vcx/tests/test_pgsql_plugin.rs:85]: Again, just voicing my opinions, feel free to leave it as is.

File: aries_vcx/tests/pgwallet/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mirgee) [reply aries_vcx/tests/pgwallet/mod.rs:2]: Still, it seems to mix postgres-specific and general storage code. Just my opinion, leave it as it is if you want.

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 5 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs:116]: Yes, so we should start moving :-) I think we can do that as part of `libvcx ---> aries-vcx` integration test migration

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 4 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs:85]: Removed logs

**Review by mirgee — APPROVED** [review]

File: aries_vcx/tests/test_pgsql_plugin.rs (hunk 3 above)
Comment (Patrik-Stas) [reply aries_vcx/tests/test_pgsql_plugin.rs]: I did this in addition to whole bunch of additional refactoring and simplification.

File: aries_vcx/tests/pgwallet/wallet_plugin_loader.rs [hunk 10]
```diff
@@ -5,12 +5,52 @@ use serde_json::Value;
...
+        .map(libloading::Library::from)
+}
+
+#[cfg(any(not(unix), not(test)))]
+pub fn load_lib(library: &str) -> libloading::Result<libloading::Library> {
+    libloading::Library::new(library)
+}
+
+
+
```
Comment (mirgee) [inline aries_vcx/tests/pgwallet/wallet_plugin_loader.rs:24]: Too many blank lines

File: aries_vcx/tests/pgwallet/wallet_plugin_loader.rs [hunk 11]
```diff
@@ -5,12 +5,52 @@ use serde_json::Value;
...
+    // based on value of xtype and OS
+    pub plugin_init_function: Option<String>,
+    // Wallet storage config for agents wallets
+    pub config: String,
+    // Wallet storage credentials for agents wallets
+    pub credentials: String,
+}
+
+pub fn init_wallet_plugin(plugin_init_config: &PluginInitConfig) -> Result<(), String> {
+    info!("_init_wallet >>> wallet_storage_config:\n{}", serde_json::to_string(plugin_init_config).unwrap());
```
Comment (mirgee) [inline aries_vcx/tests/pgwallet/wallet_plugin_loader.rs:41]: incorrect log + newline is not consistent with other logs + do we have log both entry and exit at info level?

**Review by mirgee — CHANGES_REQUESTED** [review]

**Review by mirgee — APPROVED** [review]
